guide/src/free-threading.md

```diff
@@ -0,0 +1,156 @@
+`multiprocessing` to achieve a speedup for some algorithm written in Python,
+free-threading will likely allow the use of Python threads instead for the same
+workflow.
+
+PyO3's support for free-threaded Python will enable authoring native Python
+extensions that are thread-safe by construction, with much stronger safety
+guarantees than C extensions. Our goal is to enable ["fearless
+concurrency"](https://doc.rust-lang.org/book/ch16-00-concurrency.html) in the
+native Python runtime by building on the rust `Send` and `Sync` traits.
+
+If you want more background on free-threaded Python in general, see the [what's
+new](https://docs.python.org/3.13/whatsnew/3.13.html#whatsnew313-free-threaded-cpython)
+entry in the CPython docs, the [HOWTO
+guide](https://docs.python.org/3.13/howto/free-threading-extensions.html#freethreading-extensions-howto)
+for porting C extensions, and [PEP 703](https://peps.python.org/pep-0703/),
+which provides the technical background for the free-threading implementation in
+CPython.
+
+This document provides advice for porting rust code using PyO3 to run under
+free-threaded Python. While many simple PyO3 uses, like defining an immutable
+Python class, will likely work "out of the box", there are currently some
+limitations.
+
+## Many symbols exposed by PyO3 have `GIL` in the name
+
+We are aware that there are some naming issues in the PyO3 API that make it
+awkward to think about a runtime environment where there is no GIL. We plan to
+change the names of these types to de-emphasize the role of the GIL in future
+versions of PyO3, but for now you should remember that the use of the term `GIL`
+in functions and types like `with_gil` and `GILOnceCell` is historical.
+
+Instead, you can think about whether or not a rust thread is attached to a
+Python **thread state**. See [PEP
+703](https://peps.python.org/pep-0703/#thread-states) for more background about
+Python thread states and status.
+
+In order to use the CPython C API in both the GIL-enabled and free-threaded
+builds of CPython, the thread calling into the C API must own an attached Python
+thread state. In the GIL-enabled build the thread that holds the GIL by
+definition is attached to a valid Python thread state, and therefore only one
+thread at a time can call into the C API.
+
+What a thread releases the GIL, the Python thread state owned by that thread is
+detached from the interpreter runtime, and it is not valid to call into the
+CPython C API.
+
+In the free-threaded build, more than one thread can simultaneously call into
+the C API, but any thread that does so must still have a reference to a valid
+attached thread state. The CPython runtime also assumes it is responsible for
+creating and destroying threads, so it is necessary to detach from the runtime
+before creating any native threads outside of the CPython runtime. In the
+GIL-enabled build, this corresponds to dropping the GIL with an `allow_threads`
+call.
```

Comment:
You're right, it's weaker than that, see https://docs.python.org/3/c-api/init.html#non-python-created-threads.

I've updated this so now the restriction I'm referring to only refers to OS threads that need to interact with Python objects, the rayon example you linked to is messing with rust strings, so no need to be attached to a Python interpreter runtime.

Reply:
Thanks for the link. That makes a lot of sense. In this case I think we're fine, because we don't allow the `Python` token to cross thread boundaries (`!Send`) and `Python::with_gil` (or `Python::attach` in the future?) will register the current thread correctly, right?

Reply:
Yup, `with_gil` is equivalent to calling `PyGILState_Ensure` and `PyGILStste_Release` before and after some C code that interacts with the C API.